File: leecode/3/longest-substring-without-repeating-characters.go
package leetcode

func lengthOfLongestSubstring(s string) int {
	if s == "" {
		return 0
	}
	mp := make(map[byte]bool)
	l, r, n, longest := 0, 0, len(s), 0
	for r < n {
		c := s[r]
		if _, ok := mp[c]; ok {
			longest = max(longest, r-l)
			for l < r && s[l] != c {
				delete(mp, s[l])
				l++
			}
			l++
		} else {
			mp[c] = true
		}
		r++
	}
	longest = max(longest, r-l)
	return longest
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: leecode/3/longest-substring-without-repeating-characters.cpp
#include <algorithm>
#include <set>
#include <iostream>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> store;
        int l = 0, r = 0, len = s.size();
        size_t count = 0;
        while (r < len) {
            char c = s[r];
            if (store.find(c) != store.end()) {
                count = max(count, store.size());
                while (l < r && s[l] != c) {
                    store.erase(s[l]);
                    l++; // ×óÖ¸ÕëÓÒÒÆ
                }
                l++;
            }
            else {
                store.insert(c);
            }
            r++;
        }
        return max(count, store.size());
    }
};